### PR TITLE
Update for Laravel 8 compatibility.

### LINKED ISSUE #960: Add support for Laravel 8
__Description of your feature__

Laravel 8 includes some very good improvements and new features, it would be nice to be able to update our projects that include hyn/multi-tenant to Laravel 8

---

__Proposed behavior__

Works with Laravel 8

__Current behavior__

Not compatible with laravel 8

__Information__

- laravel version: 8.0

### PR DESCRIPTION
**Pull request by mikebronner**
Fixes #960

### CODE REVIEW

**Comment by ArlonAntonius** [comment]
Please try supporting both versions and not only 8.0 😉

**Comment by mikebronner** [comment]
I don't believe that is possible due to the constructor method signature changes. I'll play around a bit and see if it works, then report back.

**Comment by mikebronner** [comment]
@ArlonAntonius I was able to get it to work. Tested with Laravel 7 and got all tests to pass. Had to hack around the model factory changes, but that's only in the test, so should be ok.

**Comment by bkintanar** [comment]
Will check on this tomorrow.

**Comment by ArlonAntonius** [comment]
The pipeline for this PR fails. There are some things that might need to be picked up on this PR:
- Multi Laravel Testing (can peak at t/t to see how we do it)
- Multi PHP Testing

(CC: @bkintanar )

**Comment by amine-mejri** [comment]
@bkintanar Any news on this PR?

**Comment by faytekin** [comment]
@bkintanar merge please

**Comment by bkintanar** [comment]
Sorry I haven't checked this yet. Will try to make time asap.

**Review by bkintanar — APPROVED** [review]

**Review by bkintanar — DISMISSED** [review]
tests are still failing.

**Comment by mikebronner** [comment]
Unfortuantely I don't have access to those test results, so those are corrections you would have to make. For what it's worth, I have been using my PR as a patch in Laravel 8 without issue. I believe you would also need to create separate tests for Laravel 7 and 8.

**Comment by vpratfr** [comment]
Would be great to see that PR merged. It has been a while since L8 is available and seems that the PR is working. Any idea on how to get moving forward?

Regarding the failing tests, it is expected to fail on PHP 7.2 environments as Laravel 8 requires PHP 7.3+

**Review by richardkeep — APPROVED** [review]

**Review by faytekin — APPROVED** [review]

**Comment by ArlonAntonius** [comment]
Hi @mikebronner,

I really appreciate the effort you have done for this. We would love to support Laravel 8. However, it would take us quite some time to adjust the test suite to get testing running properly on GHA. Instead, we decided that it would be fine to **not** support Laravel 7 anymore and completely move to Laravel 8. Would you be willing to revert the changes on this PR to your old commit that seemed to be working `c01a9cdbab2bb19151a3296219c04362af679ff0`

Once that test suite runs, we can figure out if there is anything wrong with it or if it's okay to release that as an L8 compatible version. BTW: You can check the CI yourself by simply logging with Github, should only take a few seconds and then you should be able to view the errors.

I hope you are willing to do this 🙏

**Comment by mikebronner** [comment]
Hi @ArlonAntonius, sure. I can probably get to this today or this weekend.

**Comment by mikebronner** [comment]
@ArlonAntonius @bkintanar Reverted back to Laravel 8-only and brought all tests to green.

File: ci.phpunit.xml [hunk 1]
```diff
@@ -24,15 +24,18 @@
...
         <env name="DB_USERNAME" value="root"/>
+        <env name="IN_CI" value="1"/>
+        <env name="LIMIT_UUID_LENGTH_32" value="1"/>
+        <env name="MAIL_MAILER" value="log"/>
         <env name="QUEUE_DRIVER" value="sync"/>
-        <env name="TENANCY_DATABASE_AUTO_DELETE" value="1"/>
         <env name="TENANCY_DATABASE_AUTO_DELETE_USER" value="1"/>
+        <env name="TENANCY_DATABASE_AUTO_DELETE" value="1"/>
         <env name="TENANCY_DEFAULT_HOSTNAME" value="local.testing"/>
+        <env name="TENANCY_SYSTEM_CONNECTION_NAME" value="mysql"/>
```
Comment (ArlonAntonius) [inline ci.phpunit.xml:39]: Is this really needed? As it was not needed before? 🤔

File: composer.json [hunk 2]
```diff
@@ -23,18 +23,19 @@
...
-        "laravel/framework": "^7.0"
+        "laravel/framework": "^8.0"
     },
     "require-dev": {
-        "phpunit/phpunit": "^8.0",
-        "symfony/dom-crawler": "~3.1",
-        "laravel/laravel": "^7.0",
+        "fzaninotto/faker": "^1.8",
+        "laravel/laravel": "^8.0",
+        "laravel/legacy-factories": "^1.0",
```
Comment (ArlonAntonius) [inline composer.json]: You add Legacy Factories, but you fixed a factory? Would it be worth it to fix all factories?

**Review by ArlonAntonius — APPROVED** [review]
Looks good other than that. Can someone double check?
(CC: @bkintanar @fletch3555 )

File: composer.json (hunk 2 above)
Comment (mikebronner) [reply composer.json]: 👍 I can go in and update all factories, no problem.

File: ci.phpunit.xml (hunk 1 above)
Comment (mikebronner) [reply ci.phpunit.xml:39]: I can try without it. Is it configured in the CircleCI env settings for the project?

Comment (ArlonAntonius) [reply ci.phpunit.xml:39]: Nope, but if it was not needed before, then it should not be needed now? :)

File: composer.json (hunk 2 above)
Comment (ArlonAntonius) [reply composer.json]: Would really appreciate it ❤️

**Comment by mikebronner** [comment]
@ArlonAntonius I removed the composer dependency for legacy factories. I didn't find any actual legacy factories remaining in the code. The testing env variable is required for the tests to run. Without it I was unable to run them.

The one failed test seems to be an issue with the test database connection. Perhaps rerunning the test fixes it, if its a fluke, not really sure what the cause might be.

**Comment by ArlonAntonius** [comment]
@sniper7kills Can you update the docs with a new version and simply change the requirements to reflect that L8 is supported and PHP 7.3 or higher?

**Comment by mikebronner** [comment]
Hi @ArlonAntonius, updated readme. Should hopefully be good to go? Let me know if you need anything else. Thanks! (Assuming the sniper7kills reference was in error and you meant me?)

**Comment by ArlonAntonius** [comment]
Hi @mikebronner, nope, really meant @sniper7kills , meant to update the actual docs at tenancy/docs

File: src/Traits/MutatesMigrationCommands.php [hunk 3]
```diff
@@ -14,10 +14,11 @@
 
 namespace Hyn\Tenancy\Traits;
 
-use Hyn\Tenancy\Contracts\Repositories\WebsiteRepository;
+use InvalidArgumentException;
 use Hyn\Tenancy\Database\Connection;
+use Illuminate\Contracts\Events\Dispatcher;
 use Illuminate\Database\Migrations\Migrator;
-use InvalidArgumentException;
+use Hyn\Tenancy\Contracts\Repositories\WebsiteRepository;
```
Comment (fletch3555) [inline src/Traits/MutatesMigrationCommands.php:21]: In the future, please try not to commit superfluous changes like this.  Makes it very difficult to review.

**Review by fletch3555 — APPROVED** [review]

**Pull request merged**